refactor(cli): support for `.cts`、`.mts` and follows `type: module`

molvqingtai opened the pull request:
## Related issue

Enhanced configuration files, added support for `.cts` and `.mts`, and fixed the problem of not following the user `type: module`.

## summary

`cli` will automatically register `loader` for `.ts | .cts` files, for `.ts` files marked with `type: nodule` or `.mts` will be considered as ESM and need to use `cli` to specify the loader.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

rspack.config.ts:
```ts
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

export default {
  mode: 'production',
  entry: path.resolve(__dirname, 'src/index.ts'),
  output: {
    path: path.resolve(__dirname, 'dist')
  }
}

```
package.json
```json
{
  "type": "module",
  "scripts": {
    "build": "node --loader ts-node/esm rspack build -c rspack.config.ts",
  },
  "devDependencies": {
    "@types/node": "^20.3.1",
    "ts-node": "^10.9.1",
    "typescript": "^5.1.3"
  }
}

```
tsconfig.json
```json
{
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "allowImportingTsExtensions": true,
    "noEmit": true
  }
}
```
It should work. 🎉

</details>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/web-infra-dev/rspack?pullRequest=3222) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/web-infra-dev/rspack?pullRequest=3222) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/web-infra-dev/rspack?pullRequest=3222) it.</sub>

hardfist commented:
we discuss how to support .ts in https://github.com/web-infra-dev/rspack/issues/2237 and decide to use transform strategy other than bundle strategy, so use esbuild to prebundle config is not ideal

molvqingtai commented:
> we discussed how to support .ts in #2237 and decide to use transform strategy other than bundle strategy, so use esbuild to prebundle config is not ideal

I tried to use rspack self-compilation in the code, but the output file seems to have a problem and cannot be imported or required correctly.

https://github.com/molvqingtai/rspack/blob/main/packages/rspack-cli/src/utils/loadConfig.ts#L32

hardfist commented:
> I tried to use rspack self-compilation in the code, but the output file seems to have a problem and cannot be imported or required correctly.
> 
use rspack to bundle config is not ideal which is disscussed in #2237

hardfist commented:
it's a BUG that rspack not supports "type" property of packageJson, but it should **not** be fixed by bundle config file

hardfist commented:
i think this should break into two parts type field support and mts|cts support，and we dont plan to support mts|cts

molvqingtai commented:
> we discussed how to support .ts in #2237 and decide to use transform strategy other than bundle strategy, so use esbuild to prebundle config is not ideal

Hey, I have refactored the code and updated the documentation. I used [jiti](https://github.com/unjs/jiti) without bundling packages, and the dependency issue mentioned in #2237 has been resolved.

hardfist commented:
> I tested jiti because it only runs the config file, so it's very fast, like node VM, unlike ts-node which pollutes the global scope.

sorry it seems webpack doesn't support `mts | cts | ts with type:module` either, and this implementation is too complicated and introduces tons of dependency(jiti uses babel) and edge case.
if you need to use `mts | cts | ts with type:module` i think you could use jiti on userland

alexander-akait commented:
@hardfist 

> @alexander-akait does webpack have any plan to support mts | cts | ts with type:module config?

I thought about it, but I still think that it should be on the user side, there are a lot of different tools and approaches, but I opened to discussion

hardfist commented:
https://github.com/webpack/webpack-cli/pull/3831 since webpack has supported mts|cts now, we may support it as well, @molvqingtai are you still willing to work on this?

alexander-akait commented:
@hardfist note - we don't use `jiti`, we allow to use any tools i.e. `ts-node`, `babel` and etc

molvqingtai commented:
> @hardfist note - we don't use `jiti`, we allow to use any tools i.e. `ts-node`, `babel` and etc

Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.

Use babel or other transform? I don't have a good idea at present.

alexander-akait commented:
@molvqingtai 

> Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.

Why? You can't handle all possible tools and their output

molvqingtai commented:
> @molvqingtai
> 
> > Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.
> 
> Why? You can't handle all possible tools and their output

ESM/CJS compatibility issues should not be thrown to the user. As you can imagine, when the user does not specify a loader and causes an error, it will be a bad experience to search the document for a solution.

molvqingtai commented:
> > @molvqingtai
> > > Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.
> > 
> > 
> > Why? You can't handle all possible tools and their output
> 
> ESM/CJS compatibility issues should not be thrown to the user. As you can imagine, when the user does not specify a loader and causes an error, it will be a bad experience to search the document for a solution.



> 
Using `jiit` is currently the simplest solution while maintaining a good DX. As for edge cases, both nuxt (https://github.com/nuxt/vite/pull/201) and tailwindcss(https://github.com/tailwindlabs/tailwindcss/pull/10785) are `jiti` to load configuration files. I believe this solution is stable enough.

alexander-akait commented:
@molvqingtai I strongly disagree, but I will not argue, we already tried to do it and faced with a lot of problems when new things appear (syntax/different outputs/just features like swc for `ts-node`/ES modules), why I should use babel (and dependecies) when I want to use swc. Setting up the runtime environment is just the task of the developer and no one else, you can use `jiti` under the hood if **nothing** was configured (or no options/setting provided), but always using and not allowing customization is a pretty bad idea, you can of course not take someone else's experience and go through the same problems again, your choice

alexander-akait commented:
And yes, don't forget about performance, babel is not fast

molvqingtai commented:
> And yes, don't forget about performance, babel is not fast


Of course, you can use swc. https://github.com/tailwindlabs/tailwindcss/pull/10785/commits/f1614c8d9ef50d507e271a30de6213cbaa932f0f

hardfist commented:
> > @hardfist note - we don't use `jiti`, we allow to use any tools i.e. `ts-node`, `babel` and etc
> 
> Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.
> 
> Use babel or other transform? I don't have a good idea at present.

since esm is very tricky and contains tons of edge case，i dont think let advanced user to handle esm-loader itself is a bad idea，@xiaoxiangmoe mentions jiti has some tricky design and may contain some edge case

hardfist commented:
i do agree jiti is simple and can handle 80 percent cases，but jiti has  many options which we cant expose to users and i do notice it has some lockin design which will break when javascript has new syntax or nodes invents new esm only api

molvqingtai commented:
> i do agree jiti is simple and can handle 80 percent cases，but jiti has many options which we cant expose to users and i do notice it has some lockin design which will break when javascript has new syntax or nodejs invents new esm only api

I agree with you. Can we retain the ability to customize loader and use jiit to load when no loader is specified? This ensures that the ESM configuration file can be used out of the box, and when jiti does not follow the new syntax, or some other edge cases, users can customize the loader processing.

xiaoxiangmoe commented:
I don't think using jiti is a good idea because it has some edge case. For example:

### TypeScript commonjs syntax

```ts
// file: rspack.config.ts
import * as path from "path";

const config = {
	mode: "production",
	entry: "./main.ts",
	output: {
		path: path.resolve(__dirname, "dist"),
		filename: "foo.bundle.js"
	}
};

export = config;// this stand for module.exports = config in commonjs
```

This is the usage that rspack provided now. But jiti can't handle it correctly. This will break users.

### Abusing cjs and esm

import.meta.url exist in cts
```ts
console.log(import.meta.url) // file:///Users/me/ts-node-playground/a.cts
```
__filename exist in mts

```ts
console.log(__filename) // /Users/me/ts-node-playground/a.mts
```

If we provide these feature. Should we also provide this in `.js`, `.cjs` and `.mjs`?

hardfist commented:
> To summarize the issues listed, in order to be consistent with webpack (important), if there is no better suggestion, I will delete jiit and adopt webpack's solution.

thank you for your understanding, the current wepback's way is a right(contains less edge case) but not convenient way and we may iterate to improve the DX of this in the future

hardfist commented:
Sometimes`` it seems to me that the most ideal for configurations was pure JS with jsdocs comments for types - no extra loaders, no extra transpilation, no extra overhead.

i prefer this way too，and that is why all of our examples written in this way

molvqingtai commented:
@hardfist @alexander-akait 
The refactoring is complete and the description has been updated

hardfist commented on packages/rspack-cli/package.json:
```diff
@@ -26,27 +26,23 @@
+    "interpret": "^3.1.1",
+    "rechoir": "^0.8.0",
     "semver": "6.3.0",
+    "type-fest": "^2.14.0",
```
this dependency is unnecessary we can write our own type

molvqingtai replied:
> this dependency is unnecessary we can write our own type

Looking at the type-fest source code, packageJson actually has a large number of type definitions. Is it necessary to copy the complete type definition, or simply write it as {"type": "module"}

hardfist replied:
write type: module is enough for our case

hardfist commented on packages/rspack-cli/src/utils/crossImport.ts:
```diff
@@ -0,0 +1,17 @@
+): Promise<T> => {
+	if (isEsmFile(path, cwd)) {
+		const url = pathToFileURL(path).href;
+		const { default: config } = await import(url);
```
this will crash node in jest test due to https://github.com/nodejs/node/issues/35889, and I'm not sure how webpack-cli workaround it, @alexander-akait can help me here?

alexander-akait replied:
We spawn a new process for the each test in CLI integration tests, so we don't have `jest` in scope, to collect coverage we use `nyc` (yeah, it is deprecated, but works).

hardfist replied:
i dont get it why nyc can workaround this problem here since it still runs jest  and it seems it doesnt enable vm-modules，how can jest run dynamic import without vm-moduels enabled？

alexander-akait replied:
`nyc` is not a workaround, the main workaround - we spawn a new process inside `it`, i.e. (pseudo code):
```
it("serve command", () => { 
  spawnNewProcess(__dirname, ["serve"]); 
});
```

So we don't have `jest` inside a new spawed process, it also allows you to test as if the user has run it - no polluting, just a fresh process.

`nyc` is a workaround to get coverage after testing, because we don't have `jest` inside a new spawned process we can't collect lines, so here we use `nyc` inside a new spawned process and collect coverage (but we collect it only for CI)

hardfist commented:
@molvqingtai thank you for your hard work and detailed test, the overall work LGTM

hardfist commented on packages/rspack-cli/tests/build/config/moonrepo/rspack.config.ts:
```diff
@@ -1,6 +1,6 @@
-// @ts-ignore
+// @ts-ignore: Because dynamically create {"type": "module"}
```
you can commit node_modules to git for test

hardfist commented:
@molvqingtai  I'll keep fixing the left issue to get this pr merged ASAP if you don't mind

hardfist commented:
closes #3609

molvqingtai commented:
> @molvqingtai I'll keep fixing the left issue to get this pr merged ASAP if you don't mind

Of course

hardfist commented:
!canary

hardfist commented:
@molvqingtai thank you for your great work
